src/main/java/org/osc/manager/ism/api/IsmSecurityGroupInterfaceApi.java

```diff
@@ -40,167 +41,210 @@
+        if (dev == null) {
+            String msg = String.format("Cannot find the device id: %s\n", deviceId);
+            LOGGER.error(msg);
+            throw new IllegalArgumentException(msg);
+        }
+        String existingSGIId = findSecurityGroupInterfaceByName(sgiElement.getName());
+        if (existingSGIId != null) {
+            return existingSGIId;
+        }
+
+        Set<PolicyEntity> policies = getPoliciesById(sgiElement);
+        return this.txControl.required(new Callable<String>() {
+            @Override
+            public String call() throws Exception {
+
+                SecurityGroupEntity sg = getSecurityGroupbyId(deviceId, sgiElement.getManagerSecurityGroupId());
+                SecurityGroupInterfaceEntity newSGI = new SecurityGroupInterfaceEntity(sgiElement.getName(), policies,
+                        sgiElement.getTag(), sg);
+                newSGI.setSgiDevice(dev);
+                IsmSecurityGroupInterfaceApi.this.em.persist(newSGI);
+                return newSGI.getSecurityGroupInterfaceId();
+            }
+        });
+    }
+
+    @Override
+    public void updateSecurityGroupInterface(SecurityGroupInterfaceElement sgiElement) throws Exception {
+        updateSecurityGroupInterface(this.vs.getMgrId(), sgiElement);
+    }
+
+    public void updateSecurityGroupInterface(String deviceId, SecurityGroupInterfaceElement sgiElement)
+            throws Exception {
+
+        SecurityGroupInterfaceEntity existingSgi = getSecurityGroupInterface(deviceId,
+                sgiElement.getManagerSecurityGroupInterfaceId());
+        if (existingSgi == null) {
+            String message = String.format(SGI_NOT_FOUND_MESSAGE, sgiElement.getManagerSecurityGroupInterfaceId());
+            throw new Exception(message);
+        }
+        existingSgi.setName(sgiElement.getName());
+        existingSgi.setPolicies(getPoliciesById(sgiElement));
+        existingSgi.setSecurityGroup(getSecurityGroupbyId(deviceId, sgiElement.getManagerSecurityGroupId()));
+        existingSgi.setTag(sgiElement.getTag());
+        this.txControl.required(new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+                IsmSecurityGroupInterfaceApi.this.em.merge(existingSgi);
+                return null;
+            }
+        });
+    }
+
+    @Override
+    public void deleteSecurityGroupInterface(String id) throws Exception {
+        deleteSecurityGroupInterface(this.vs.getMgrId(), id);
+    }
+
+    public void deleteSecurityGroupInterface(String deviceId, String id) throws Exception {
+        this.txControl.required(new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+                SecurityGroupInterfaceEntity existingSgi = getSecurityGroupInterface(deviceId, id);
+                if (existingSgi == null) {
+                    LOGGER.info(String.format(SGI_NOT_FOUND_MESSAGE, id));
```

Comment:
extra ln

Reply:
between the two methods I have added the extra line for readability. This should still be fine.